### ISSUE TITLE
🚀 Feature Request: Do TypeScript type checking on `wrangler deploy` 

### ISSUE THREAD

**Issue by ramesaliyev** [cloudflare/workers-sdk#6321]
### Describe the solution

Hello!

Wrangler uses [esbuild](https://esbuild.github.io) for bundling and since **esbuild does not do any type checking** (see evanw/esbuild#1631) no errors are produced if there is something wrong. This kinda defeats the purpose.

From [esbuild's documentation on TypeScript](https://esbuild.github.io/content-types/#typescript):

> However, esbuild _does not_ do any type checking so you will still need to run `tsc -noEmit` in parallel with esbuild to check types. This is not something esbuild does itself.

---

I believe [`tsc --noEmit`](https://www.typescriptlang.org/tsconfig/#noEmit) or something else needs to be a part of `wrangler deploy`. Or at least this information, and a solution, should be added to somewhere in documentation. Maybe somewhere in [Bundling](https://developers.cloudflare.com/workers/wrangler/bundling) .

Because I only noticed this by chance. There may be some people out there who are not aware of this. 

---

I'll post my current solution in case someone else needs it.

So running `tsc --noEmit` directly on repo root fails with following logs even when said modules are installed.

```bash
error TS2688: Cannot find type definition file for '@cloudflare/kv-asset-handler'.
  The file is in the program because:
    Entry point of type library '@cloudflare/kv-asset-handler' specified in compilerOptions

  tsconfig.json:38:4
    38    "@cloudflare/kv-asset-handler"
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    File is entry point of type library specified here.

error TS2688: Cannot find type definition file for '@cloudflare/workers-types/2023-07-01'.
  The file is in the program because:
    Entry point of type library '@cloudflare/workers-types/2023-07-01' specified in compilerOptions

  tsconfig.json:37:4
    37    "@cloudflare/workers-types/2023-07-01",
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    File is entry point of type library specified here.


Found 2 errors.
```

I had to create another tsconfig file and run it like `tsc --project tsconfig.errcheck.json` to manage type checking. Using this as a [custom build](https://developers.cloudflare.com/workers/wrangler/custom-builds/) step is working for me now.

```json
{
  "compilerOptions": {
    "target": "esnext",
    "skipLibCheck": true,
    "noEmit": true,
    "moduleResolution": "bundler",
    "typeRoots": [
      "./types"
    ]
  },
  "include": [
    "worker-configuration.d.ts",
    "src/**/*.ts",
    "types/**/*.d.ts"
  ]
}
```

Thanks!

**Comment by threepointone** [comment]
I'd push back against this. It's not wrangler's purview to enforce type checking (should it fail? warn? should it support all of tsc's flags? etc etc.) That said, maybe the templates generated could include an npm script entry that calls tsc --no-Emit? dunno. 

That said, the issue you pointed out with kv-asset-handler seems not directly related. Could you expand on that more? typechecking is failing in the workers-sdk repo for you? Or in a generated app?

**Comment by ramesaliyev** [comment]
Yeah maybe my wording is a little bit off here. What I actually meant is that when I generate a worker app with typescript by using `npm create cloudflare`, the generated app should do type checking before deployment. This can be achieved adding custom builds entry to generated `wrangler.toml` which performs `tsc --noEmit` like you said. 

About the second issue, just ignore it, it was my fault, edited the original issue.

**Comment by petebacondarwin** [comment]
We are considering adding a `wrangler check` command that would do type checking (among other things). We could then consider whether a C3 template would run this before `wrangler deploy` if the user is actually using TypeScript...

**Comment by penalosa** [comment]
Apologies for the delay here! We don't intend to add support for running `tsc` as part of `wrangler deploy` (https://github.com/cloudflare/workers-sdk/issues/6321#issuecomment-2249884045), and so I'm going to close this for now.
